File: 最長遞增子序列.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

int binarySearch(int *array, int arraySize, int key) {
    int left = 0, right = arraySize - 1;
    while(left <= right) {
        int mid = left + ((right - left) >> 1);
        if(array[mid] < key)
            left = mid + 1;
        else
            right = mid - 1;
    }
    return left;
}

int lengthOfLIS(int *array, int arraySize) {
    int *LIS = NULL, length = 0;
    for(int i = 0; i < arraySize; ++i) {
        int pos = binarySearch(LIS, length, array[i]);
        if(pos == length)
            LIS = (int*)realloc(LIS, ++length * sizeof(int));
        LIS[pos] = array[i];
    }
    free(LIS);
    return length;
}

int main() {
    int *array = NULL, arraySize = 0, n;
    while(scanf("%d", &n) && n != 0) {
        array = (int*)realloc(array, (arraySize + 1) * sizeof(int));
        array[arraySize++] = n;
    }
    fputs("Length of LIS of [", stdout);
    bool firstNum = true;
    for(int i = 0; i < arraySize; ++i) {
        if(!firstNum)
            fputs(", ", stdout);
        printf("%d", array[i]);
        firstNum = false;
    }
    printf("] is %d\n", lengthOfLIS(array, arraySize));
    free(array);
    return 0;
}

File: 最長遞增子序列.java
import java.io.*;
import java.util.*;

public class Main {
	public static void main(String[] args) throws IOException {
	    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StreamTokenizer st = new StreamTokenizer(br);
        ArrayList<Integer> array = new ArrayList<>();
        int n;
        while(st.nextToken() == StreamTokenizer.TT_NUMBER && (n = (int)st.nval) != 0)
            array.add(n);
        System.out.println("Length of LIS of " + array.toString() + " is " + lengthOfLIS(array));
	}
	
	static int binarySearch(ArrayList<Integer> array, int key) {
        int left = 0, right = array.size() - 1;
        while(left <= right) {
            int mid = left + (right - left >> 1);
            if(array.get(mid) < key)
                left = mid + 1;
            else
                right = mid - 1;
        }
        return left;
    }
    
    static int lengthOfLIS(ArrayList<Integer> array) {
        ArrayList<Integer> LIS = new ArrayList<>();
        int length = 0;
        for(Integer i : array) {
            int pos = binarySearch(LIS, i);
            if(pos == LIS.size())
                LIS.add(i);
            else
                LIS.set(pos, i);
        }
        return LIS.size();
    }
}
